SPI API update

tbursztyka opened the pull request:
Only DW driver got an update for now. The legacy API is still the default.

Commit:
api: New SPI API

Such API improves many aspects of the former API by reducing the number
of function, allowing more buffer flexibility etc... This leads in
better memory usag and performance as well.

However, as this will take sometime to get into use, the former API is
still present and is the one enabled by default.

Jira: ZEP-852
Jira: ZEP-287
Jira: ZEP-1725

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Apply syntax rules on DW drivers

One liners if/for/while statements still need {}
(and line break are cheap for clarity).

Aligning parameters properly.

Also, removing __func__ usage from SYS_LOG_* as these macros already put
it internally.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86/quark_se: Let the possibility to use native SPI driver

Set the right definitions to be able to use SPI_DW instead of SPI_QMSI.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples: Add an SPI driver test application

This application intents to test an SPI driver by looping over MISO/MOSI
line where the controller will then send data to itself.

It will test various buffer tx/rx schemes.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Adapt DW driver to new SPI API

Introducing as well a generic driver helper for CS gpio control and
buffer management.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api: Add asynchronous call support to SPI API

Adding a struct k_poll_signal parameter to driver's API unique
exposed function.

If not NULL, the call will be handled as asynchronous and will
return right after the transaction has started, on the contrary
of current logic where is waits for the transaction to finish
(= synchronous).

In order to save stack, let's move the device pointer to struct
spi_config. So the call is still at a maximum of 4 parameters.

Adapting spi_dw.c and spi driver sample to the change so it still
builts.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Add reentrance support to DW driver in a generic manner

Let's improve common SPI driver context by adding a lock and generic
function to get/release it.

It's statically initialized to save a bit of ROM.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Handle synchronous calls in a generic manner in DW driver

All SPI drivers have this same way to handle synchronous call, thus
let's generalize it in struct spi_context, with a relevant API and apply
the change into SPI DW driver.

spi_context API will prove to be useful once asynchronous call will be
handled as well, through the same completion functions used now only for
synchronous call. It will be transparent for the driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Add support for async call in DW driver in a generic way

All is done through the generic spi_context driver's API as it will be
generic to all SPI drivers.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/spi: Add an asynchronous call test

This gives a quicke example on how to use SPI asynchronous calls with
kernel's k_poll API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Add 2 specific control bits for special use cases.

SPI_HOLD_ON_CS can be used to ask the SPI device to keep CS on, after
the transaction. And this undefinitely, until another config is used.
This will inhibate the gpio cs delay, if any. This might be useful when
doing consecutive calls on one slave without releasing the CS.

SPI_LOCK_ON is to be used with caution as it will keep the SPI device
locked for the current config being used after each transaction. This
can be necessary if one needs to do consecutive calls on a slave without
any olher caller to interfere.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/spi: Handle ressource locking and release in DW driver

Again this is made as generic as possible through driver's
spi_context API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/spi: Add a quick test for spi_release() API function

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

mbolivar commented on include/spi.h:
```diff
@@ -27,173 +37,182 @@
- * on the same line.
+ * @param dev is a valid pointer to an actual SPI device
+ * @param config Pointer to a valid spi_config structure instance.
+ * @param tx_bufs Memory buffer(s) where data to be sent originates from.
```
Looks like a good improvement.

Are tx_bufs and rx_bufs null-terminated? If so, could this be documented?

tbursztyka replied:
It's indeed missing, I'll do that above at spi_buf declaration.

mbolivar commented on include/spi.h:
```diff
@@ -27,173 +37,182 @@
-typedef int (*spi_api_io)(struct device *dev,
-			  const void *tx_buf, u32_t tx_buf_len,
-			  void *rx_buf, u32_t rx_buf_len);
+typedef int (*spi_api_io)(struct device	*dev,
```
It seems a bit harder to tease apart the errors due to misconfiguration from I/O errors here than in the old API. Will you specify which errno values correspond to which?

In general, what is the rationale for removing the separate configuration routine?

This seems to add some overhead, as the SPI controller must re-check the configuration each time its spi_api_io is called, rather than being able to kick off new I/O with an already established configuration. This seems like it could decrease throughput in back-to-back transactions.

mbolivar replied:
Ah, I see now that you left a comment in the PR about the configuration versus I/O errors.

tbursztyka replied:
Take a look at the spi_dw.c in next patch: if given config is same as the driver has been using previously, the controller won't be reconfigured and transaction will proceed right away.

Error code would be -EINVAL (mis-configuration), -ENOTSUP (requesting a mode or else not supported), -EIO for any transaction error.

mbolivar commented on include/spi.h:
```diff
@@ -138,10 +142,10 @@ struct spi_buf {
 			  const struct spi_buf **tx_bufs,
-			  struct spi_buf **rx_bufs);
+			  struct spi_buf **rx_bufs,
+			  struct k_poll_signal *async);
```
Nice :+1:

Commit:
drivers/spi: Handle SPI_HOLD_ON_CS in spi context through gpio

If CS is controlled over GPIO, it will be possible to keep the slave up
and running (though no transaction will be going on) using this
configuration bit.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
api/spi: Add EEPROM configuration bit

EEPROM read mode is a specific mode where the controller will TX a
command to the slave, and once done, will read as many bytes requested.
The gain relies in the controller generating all necessary dummy bytes
by itself to read data the from slave, it will only generate RX
interrupts. Thus reducing CPU work.
Obviously TX and RX buffers should be relevantly provided by the user.

If not supported by the controller, the driver can still work (it will
have to generate the dummy bytes) and thus -EINVAL should not be
returned for that configuration bit.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

mnkp commented on include/spi.h:
```diff
@@ -27,173 +37,184 @@
-typedef int (*spi_api_io)(struct device *dev,
-			  const void *tx_buf, u32_t tx_buf_len,
-			  void *rx_buf, u32_t rx_buf_len);
+typedef int (*spi_api_io)(struct device	*dev,
```
We should better use the name spi_api_transceive_t

tbursztyka commented on include/spi.h:
```diff
@@ -27,173 +37,184 @@
+#define SPI_WORD_SIZE_GET(_operation_)		\
+	(((_operation_) & SPI_WORD_SIZE_MASK) >> 5)
+
+#define SPI_WORD(_word_size_)			\
```
Actually, in the early days of Zephyr, we had a discussion about it. It got decided to move on with bit mask at that time, and well I decided to keep it that way.
Thing is, for these configuration bits, it will be most of the time built-time computed so in the end it will end up as a static number assigned to struct spi_config's operation attribute.

tbursztyka commented on include/spi.h:
```diff
@@ -27,173 +37,184 @@
+/**
+ * @brief SPI word size
+ */
+#define SPI_WORD_SIZE_MASK	(0x3F << 5)
```
True, I actually used that way in my spi_update_wip branch (it's about supporting some features in dw driver).

tbursztyka commented on include/spi.h:
```diff
@@ -27,173 +37,184 @@
-typedef int (*spi_api_slave_select)(struct device *dev, u32_t slave);
+struct spi_buf {
+	void *buf;
+	u32_t len;
```
indeed, should be the same actually as zephyr currently only support 32bits arch.

tbursztyka commented on include/spi.h:
```diff
@@ -27,173 +37,184 @@
+#define SPI_TRANSFER_LSB	BIT(4)
 
-#define SPI_TRANSFER_MASK	(0x8)
+#define SPI_TRANSFER_MASK	(0x10)
```
For 1 bit setting actually, I don't think we need any mask.

galak commented:
are we addressing slave mode in the new API?

pfalcon commented:
> are we addressing slave mode in the new API?

Is that even possible? Slave mode support seem so wildly different for different SoCs, so one may wonder if it's possible to come up with any (generic) common API. I'd love to read a survey of slave mode capabilities which comes to some common ground...

tbursztyka commented:
Only the slave bit for configuration is there for now (for runtime controller reconfiguration, when supported obviously). But I am still trying to find the right way to support slave operation properly, as it has to reply at first word received (and that is not going to be made in ISR mode, as building the reply may take a good amount of stack). All that has to be considered to get the easiest, yet flexible, API.

mrrosen commented:
@tbursztyka 

This API has included my most important ask for holding the CS after a single transaction. One thing I might suggest is that passing in NULL for the k_poll_signal async argument of the asynchronous functions shouldnt result in the function being blocking but just not having a way of knowing when the transaction (because the user doesnt care). True that you can just pass in a k_poll_signal that you never check again but that might be difficult to free that resource later. Just another call to the spi driver will result in a block under the transaction is complete anyway, correct?

Second, I think this is a very specific case that is fixed via the explicit GPIO device for the CS in some ways; but that weird SD card requirement of running the SCK without CS asserted doesnt seem to be possible in the current API (unless passing in a NULL for the spi_cs_control means the drivers will not try to control CS for that transaction, in which case, nevermind!).

I do agree that the name operation is a bit confusing and I think flags would be a better name.

tbursztyka commented:
The k_poll_signal being NULL idea means the driver would need to expose a different function to do this async stuff. Well, does not seem to add up that much overhead, and I like the idea of throwing a call and forgetting about it. But yes, other competing call will block as long as this call is running. And it will simplify things so that if CONFIG_POLL is unset, the async function won't be exposed at all (and there won't be a useless parameter as now).

Yep, passing a NULL as spi_cs_control is fully nominal. It's either/or, nothing forced here.

m-kru commented on drivers/spi/spi_context.h:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	ctx->tx_len--;
+	if (!ctx->tx_len) {
```
Can you please document these lines. They are not so obvious and it may help you to avoid questions from that are not as experienced as you are.

mbolivar commented:
@nashif @galak @MaureenHelm @tbursztyka: will Zephyr still be accepting master mode drivers that use the old API, or should out of tree drivers be re-worked to use the new API before submission?

We have an STM32F4 SPI driver that I'm going to be getting ready for submission soon (https://github.com/Linaro/zephyr/commit/f4668aa8f1ac41f7fe9172ea729fbf3f78c53ba9), and I'd like to know what will be needed.
